src/components/ha-web-rtc-player.ts

```diff
@@ -119,98 +171,217 @@ class HaWebRtcPlayer extends LitElement {
+        this._peerConnection!.onicegatheringstatechange = (ev: Event) => {
+          const iceGatheringState = (ev.target as RTCPeerConnection)
+            .iceGatheringState;
+          if (iceGatheringState === "complete") {
+            resolve();
+          }
+
+          console.timeLog(
+            "WebRTC",
+            "Ice gathering state changed",
+            iceGatheringState
+          );
+        };
       });
-    });
-    await iceResolver;
 
-    console.timeLog("WebRTC", "end iceResolver", candidates);
+      if (!this._peerConnection || !this.entityid) {
+        return;
+      }
+    }
+
+    while (this._candidatesList.length) {
+      const candidate = this._candidatesList.pop();
+      if (candidate) {
+        candidates += `a=${this._candidatesList.pop()}\r\n`;
+      }
+    }
 
     const offer_sdp = offer.sdp! + candidates;
 
-    let webRtcAnswer: WebRtcAnswer;
+    console.timeLog("WebRTC", "start webRtcOffer", offer_sdp);
+
     try {
-      console.timeLog("WebRTC", "start WebRTCOffer", offer_sdp);
-      webRtcAnswer = await handleWebRtcOffer(
-        this.hass,
-        this.entityid,
-        offer_sdp
+      this._unsub = webRtcOffer(this.hass, this.entityid, offer_sdp, (event) =>
+        this._handleOfferEvent(event)
       );
-      console.timeLog("WebRTC", "end webRtcOffer", webRtcAnswer);
     } catch (err: any) {
       this._error = "Failed to start WebRTC stream: " + err.message;
-      peerConnection.close();
+      this._cleanUp();
+    }
+  };
+
+  private _iceConnectionStateChanged = () => {
+    console.timeLog(
+      "WebRTC",
+      "ice connection state change",
+      this._peerConnection?.iceConnectionState
+    );
+    if (this._peerConnection?.iceConnectionState === "failed") {
+      this._peerConnection.restartIce();
+    }
+  };
+
+  private async _handleOfferEvent(event: WebRtcOfferEvent) {
+    if (!this.entityid) {
       return;
     }
+    if (event.type === "session_id") {
+      this._sessionId = event.session_id;
+      this._candidatesList.forEach((candidate) =>
+        addWebRtcCandidate(
+          this.hass,
+          this.entityid!,
+          event.session_id,
+          candidate
+        )
+      );
+      this._candidatesList = [];
+    }
+    if (event.type === "answer") {
+      console.timeLog("WebRTC", "answer", event.answer);
 
-    // Setup callbacks to render remote stream once media tracks are discovered.
-    const remoteStream = new MediaStream();
-    peerConnection.addEventListener("track", (event) => {
-      console.timeLog("WebRTC", "track", event);
-      remoteStream.addTrack(event.track);
-      this._videoEl.srcObject = remoteStream;
-    });
-    this._remoteStream = remoteStream;
+      this._handleAnswer(event);
+    }
+    if (event.type === "candidate") {
+      console.timeLog("WebRTC", "remote ice candidate", event.candidate);
+
+      try {
+        await this._peerConnection?.addIceCandidate(
```

Comment:
What should happen if `_peerConnection` is undefined? Do we just do nothing with a candidate? can this even happen?

Reply:
Yes, we do nothing with the candidate. It could be a race where we have cleaned peerConnection already but the ws event was not processed yet. We dont care about the candidate anymore at that point.

Reply:
Yes, we do nothing with the candidate. It could be a race where we have cleaned peerConnection already but the ws event was not processed yet. We dont care about the candidate anymore at that point.

Reply:
Should we replace this by a return statement? So it's clear that we do nothing : 
```ts
if (!this._peerConnection) {
   return;
}
```